packages/ra-core/src/controller/input/useReferenceInputController.ts

```diff
@@ -84,12 +87,23 @@ const useReferenceInputController = ({
     filterToQuery,
     referenceSource = defaultReferenceSource,
     resource,
+    sort: sortOverride,
     source,
 }: Option): ReferenceInputValue => {
     const translate = useTranslate();
 
     const { pagination, setPagination } = usePaginationState({ perPage });
-    const { sort, setSort } = useSortState();
+    const { sort, setSort } = useSortState(sortOverride);
+
+    // We need to keep the current sort synchronized with the sort prop
+    // because the inner component may call the setSort too
+    useEffect(() => {
+        if (!isEqual(sort, sortOverride)) {
```

Comment:
why the `!`? I'd expect that the effect does nothing if the props are equal, and sets the sort if the props are different. That makes me think that you are probably missing a unit test.